src/full_node/address_manager_store.py

```diff
@@ -0,0 +1,231 @@
+from src.server.address_manager import (
+    AddressManager,
+    ExtendedPeerInfo,
+    NEW_BUCKET_COUNT,
+    BUCKET_SIZE,
+    NEW_BUCKETS_PER_ADDRESS,
+)
+
+import logging
+import aiosqlite
+
+log = logging.getLogger(__name__)
+
+
+class AddressManagerStore:
+    """
+    Metadata table:
+    - private key
+    - new table count
+    - tried table count
+    Nodes table:
+    * Maps entries from new/tried table to unique node ids.
+    - node_id
+    - IP, port, together with the IP, port of the source peer.
+    New table:
+    * Stores node_id, bucket for each occurrence in the new table of an entry.
+    * Once we know the buckets, we can also deduce the bucket positions.
+    Every other information, such as tried_matrix, map_addr, map_info, random_pos,
+    be deduced and it is not explicitly stored, instead it is recalculated.
+    """
+    db: aiosqlite.Connection
+
+    @classmethod
+    async def create(cls, connection):
+        self = cls()
+        self.db = connection
+        await self.db.commit()
+
+        await self.db.execute(
+            "CREATE TABLE IF NOT EXISTS peer_metadata("
+            "key text,"
+            "value text)"
+        )
+        await self.db.commit()
+
+        await self.db.execute(
+            "CREATE TABLE IF NOT EXISTS peer_nodes("
+            "node_id int,"
+            "value text)"
+        )
+        await self.db.commit()
+
+        await self.db.execute(
+            "CREATE TABLE IF NOT EXISTS peer_new_table("
+            "node_id int,"
+            "bucket int)"
+        )
+        await self.db.commit()
+        return self
+
+    async def clear(self):
+        cursor = await self.db.execute("DELETE from peer_metadata")
+        await cursor.close()
+        await self.db.commit()
+        cursor = await self.db.execute("DELETE from peer_nodes")
+        await cursor.close()
+        await self.db.commit()
+        cursor = await self.db.execute("DELETE from peer_new_table")
+        await cursor.close()
+        await self.db.commit()
+
+    async def get_metadata(self):
+        cursor = await self.db.execute("SELECT key, value from peer_metadata")
+        metadata = await cursor.fetchall()
+        await cursor.close()
+        return {key: value for key, value in metadata}
+
+    async def is_empty(self):
+        metadata = await self.get_metadata()
+        if "key" not in metadata:
+            return True
+        if (
+            "new_count" in metadata
+            and metadata["new_count"] > 0
```

Comment:
use this pattern `metadata.get("new_count", 0) > 0` instead in all these checks.

Reply:
+1